### DISCUSSION TITLE
[zodios-openapi] Per route security schemes

### DISCUSSION THREAD

**Discussion by epotter2297** [ecyrbe/zodios#186] [Ideas]
If I understand correctly, right now the only way to add a security scheme to OpenAPI is by defining `securityScheme` in the `toOpenApi()` function, which will then add a security scheme to every endpoint. However, there currently isn't a way for me to only add it to some endpoints.

As an example, if I have a ` POST /email` endpoint that requires a ReCAPTCHA Authorization header (and I don't want to use that header anywhere else), it is not possible for me to do that with the current implementation of zodios-openapi.

---
Somewhat related side-note: I'm aware that zodios-openapi is fairly early in it's life, and it's likely some of these features just haven't gotten around to being added yet. But I just wanted to add that being able to specify multiple types of authorization and name them (ie: a "jwt" Authorization header for admin routes, and a "ReCAPTCHA" Authorization header for sending emails) would be very nice to have too 🙂 

---
A solution (which could solve both of the above issues) could be to add a `makeSecuritySchemes()` function which could be used like so:
```js
const schemes = makeSecuritySchemes([
  {
    name: 'Scheme A',
    securityScheme: basicAuthScheme(),
  },
  {
    name: 'Scheme B',
    securityScheme: bearerAuthScheme(),
  },
]);

...

const api = schemes.makeApi([
  {
    method: 'get',
    path: '/users/:id',
    alias: 'getUser',
    description: 'Get a user',
    securitySchemes: ['Scheme A'],
    parameters: [
      {
        name: 'id',
        type: 'Path',
        schema: z.number(),
      },
    ],
    response: z.object({
      id: z.number(),
      name: z.string().nullable(),
    }),
  },
]);

...

const document = toOpenApi(api, {
  info: {
    title: 'Project A API',
    version: '0.0.0',
    description: 'A simple example API',
  },
  servers: [
    {
      url: '/',
    },
  ],
  securitySchemes: schemes, // would this even be necessary if we used schemes.makeApi()?
});
```
I used the same concept for how `ctx` works when creating your routes while designing this. There are probably implications of doing things this way that I haven't considered, and I'm not sure if this would be what the final API would look like, but it's a good starting point I think 🙂

**Comment by ecyrbe** [comment]
Hello,

I usually think that per endpoint scheme is overkill since you usually want to have schemes grouped. 
So is ok for openapi spec definition since they have no javascript to simplify declaration, but since we have this power i think we should take the opportunity to make things simpler.
So maybe to answer all needs, but allowing for less repeated scheme declaration would be to have something that look like this:

```ts
cons document = toMultiSchemeOpenApi({
  info: {
    title: 'Project A API',
    version: '0.0.0',
    description: 'A simple example API',
  },
  servers: [
    {
      url: '/',
    },
  ],
  securitySchemeAPIs: [
    {
      securityScheme: basicAuthScheme(),
      api: basicAuthAPI,
    },
    {
      securityScheme: bearerAuthScheme(),
      api: bearerAuthAPI,
    },
    {
        api: notSecuredAPI, // for health check, etc
    }
  ] 
});
```

This has the advantage of : 
- less repetition
- no api definition changes
- no breaking change on `toOpenApi`

what do you think ?

**Reply by epotter2297** [reply]
Yeah, I think that your solution can solve all of the same problems and be more concise / simpler.

I do think that having the ability to name your security schemes would be helpful though (could just be an extra `name` property along with `securityScheme` and `api`), because sometimes I use the same type of header for different things.

Here's an example of some swagger docs from a different API that I have, I would want to replicate this same thing with zodios-openapi
<img width="643" alt="Screen Shot 2022-10-17 at 3 20 02 PM" src="https://user-images.githubusercontent.com/65096111/196264848-d2749a79-bdfb-44db-bf23-aef1d96d0ea7.png">

**Reply by ecyrbe** [reply]
yes, fair point. completly agree.

**Comment by ecyrbe** [comment]
@epotter2297 i created an issue to track this #187

**Comment by ecyrbe** [comment]
hello @epotter2297 ,

This is now available on `v10.4.0` of `@zodios/openapi`.

You'll see that finally api looks really different than what i proposed, but it's even simpler now, since this feature is available behind a builder.

Link to [documentation](http://www.zodios.org/docs/api/openapi#swagger-ui-express)
